01_Data/src/layers/sequelize/model/Tariff/Tariffs.ts

```diff
@@ -1,32 +1,295 @@
 // Copyright Contributors to the CitrineOS Project
 //
 // SPDX-License-Identifier: Apache 2.0
+import {EnergyMixData} from "./EnergyMix";
 
-import { Namespace } from '@citrineos/base';
-import { Column, DataType, Index, Model, Table } from 'sequelize-typescript';
-import { TariffUnitEnumType } from './index';
+import {
+  assert,
+  CountryCode,
+  Currency,
+  Displaytext,
+  isCountryCode,
+  isCurrency,
+  isPartyId,
+  isTariffId,
+  isUrl,
+  Namespace,
+  PartyId,
+  TariffId,
+  TariffKey,
+  Url
+} from '@citrineos/base';
+import {Column, DataType, HasMany, Index, Model, PrimaryKey, Table} from 'sequelize-typescript';
+import {Price} from "./Price";
+import {TariffElement, TariffElementData} from "./TariffElement";
+
+export const ALL_TARIFF_TYPES = [
+  'AD_HOC_PAYMENT',
+  'PROFILE_CHEAP',
+  'PROFILE_FAST',
+  'PROFILE_GREEN',
+  'REGULAR'
+] as const;
+
+export type TariffType = typeof ALL_TARIFF_TYPES[number];
+
+function isTariffType(value: string): value is TariffType {
+  return (ALL_TARIFF_TYPES as readonly string[]).includes(value);
+}
+
+export function tariffType(value: string): TariffType {
+  assert(isTariffType(value), `Invalid tariff type: ${value}`);
+  return value;
+}
 
 @Table
-export class Tariff extends Model {
+export class Tariff extends Model implements TariffData {
+
   static readonly MODEL_NAME: string = Namespace.Tariff;
 
-  /**
-   * Fields
-   */
+  public static newInstance(data: TariffData): Tariff {
+    return Tariff.build({...data}, {include: [TariffElement]});
+  }
+
+  @PrimaryKey
```

Comment:
I understand we want a composite primary key of (tariffId, countryCode, partyId)?